test/teststl.cpp

```diff
@@ -4411,6 +4413,79 @@ class TestStl : public TestFixture {
               true);
         ASSERT_EQUALS("", errout.str());
     }
+
+    void checkMutexes() {
+        check("void f() {\n"
+              "    static std::mutex m;\n"
+              "    static std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (error) The lock guard won't unlock until the end of the program which could lead to a deadlock.\n", errout.str());
+
+        check("void f() {\n"
+              "    static std::mutex m;\n"
+              "    std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("", errout.str());
+
+        check("void f() {\n"
+              "    std::mutex m;\n"
+              "    std::lock_guard<std::mutex> g(m);\n"
+              "}\n",
+              true);
+        ASSERT_EQUALS("[test.cpp:3]: (error) The mutex is locked at the same scope as the mutex itself.\n", errout.str());
+
+        check("void g();\n"
+              "void f() {\n"
+              "    static std::mutex m;\n"
```

Comment:
maybe an additional check; I am not sure it's thread safe to have a local static mutex in a function. If the function is called twice at the same time and the mutex constructor is executed twice...

Reply:
> If the function is called twice at the same time and the mutex constructor is executed twice...

No its not, from [cppreference](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables):

> If multiple threads attempt to initialize the same static local variable concurrently, the initialization occurs exactly once (similar behavior can be obtained for arbitrary functions with std::call_once).

Reply:
thanks! :-)

Reply:
I should update our threadsafety addon so it does not warn about c++ code then.

Reply:
> I should update our threadsafety addon so it does not warn about c++ code then.

Well it can still be a threadsafety issue if a non-const reference to the static variable is returned:

```cpp
Foo& get_foo()
{
    static Foo x;
    return x;
}
```

Since the static variable could be modified across multiple threads. I dont how extensive threadsafety checks for. I assume since its an addon it is a little more opinionated.

Reply:
The plugin initially targets  pre-C++11 code where it's not guranteed to be thread-safe.